### ISSUE TITLE
A job stuck endlessly for days.

### ISSUE THREAD

**Issue by abhijitsinha** [mperham/sidekiq#2681]
Hello Mike,

We are using sidekiq pro for our app Realvolve. Here is the list of what we are using and version - 

- ruby '2.1.2'
- rails '3.2.18'
- sidekiq '3.2.1'
- sidekiq-pro '1.7.4'

We referred your blog http://www.mikeperham.com/2013/05/03/dealing-with-stuck-workers/, where you say: 

*Log into the host noted in the list and look for the process with the PID. If the PID exists, the job is probably locked up somehow.*

We see the process running. We also gave a TTIN signal and recieved a log dump.

From the dump( I have pasted below) we are unable to conclude what is the cause of the issue - Its a very simple job which receives few params and sends an email. We have never faced issue with any other job like this - is there a way to remove this job?

Any help is highly appreciated. Thanks

Here is the dump:

```
2015-11-16T16:37:50.951Z 25390 TID-ovevxfqxs INFO: Thread TID-ovew4h5gw 
2015-11-16T16:37:50.951Z 25390 TID-ovevxfqxs INFO: /home/realvolve/app/vendor/bundle/ruby/2.1.0/gems/premailer-1.8.2/lib/premailer/html_to_plain_text.rb:29:in `gsub!'
/home/realvolve/app/vendor/bundle/ruby/2.1.0/gems/premailer-1.8.2/lib/premailer/html_to_plain_text.rb:29:in `convert_to_text'
/home/realvolve/app/vendor/bundle/ruby/2.1.0/gems/premailer-1.8.2/lib/premailer/adapter/hpricot.rb:164:in `to_plain_text'
/home/realvolve/app/vendor/bundle/ruby/2.1.0/gems/premailer-rails-1.8.0/lib/premailer/rails/hook.rb:81:in `generate_text_part'
/home/realvolve/app/vendor/bundle/ruby/2.1.0/gems/premailer-rails-1.8.0/lib/premailer/rails/hook.rb:62:in `generate_alternative_part'
/home/realvolve/app/vendor/bundle/ruby/2.1.0/gems/premailer-rails-1.8.0/lib/premailer/rails/hook.rb:50:in `generate_html_part_replacement'
/home/realvolve/app/vendor/bundle/ruby/2.1.0/gems/premailer-rails-1.8.0/lib/premailer/rails/hook.rb:24:in `perform'
/home/realvolve/app/vendor/bundle/ruby/2.1.0/gems/premailer-rails-1.8.0/lib/premailer/rails/hook.rb:8:in `perform'
/home/realvolve/app/vendor/bundle/ruby/2.1.0/gems/mail-2.5.4/lib/mail/mail.rb:230:in `block in inform_interceptors'
/home/realvolve/app/vendor/bundle/ruby/2.1.0/gems/mail-2.5.4/lib/mail/mail.rb:229:in `each'
/home/realvolve/app/vendor/bundle/ruby/2.1.0/gems/mail-2.5.4/lib/mail/mail.rb:229:in `inform_interceptors'
/home/realvolve/app/vendor/bundle/ruby/2.1.0/gems/mail-2.5.4/lib/mail/message.rb:220:in `inform_interceptors'
/home/realvolve/app/vendor/bundle/ruby/2.1.0/gems/mail-2.5.4/lib/mail/message.rb:230:in `deliver'
/home/realvolve/app/lib/jobs/send_action_messages/send_email.rb:13:in `block in perform'
/home/realvolve/app/lib/jobs/send_action_messages/send_email.rb:9:in `each'
/home/realvolve/app/lib/jobs/send_action_messages/send_email.rb:9:in `perform'
/home/realvolve/app/vendor/bundle/ruby/2.1.0/gems/sidekiq-3.2.1/lib/sidekiq/processor.rb:52:in `block (2 levels) in process'
/home/realvolve/app/vendor/bundle/ruby/2.1.0/gems/sidekiq-3.2.1/lib/sidekiq/middleware/chain.rb:122:in `call'
/home/realvolve/app/vendor/bundle/ruby/2.1.0/gems/sidekiq-3.2.1/lib/sidekiq/middleware/chain.rb:122:in `block in invoke'
/home/realvolve/app/vendor/bundle/ruby/2.1.0/gems/statsd-ruby-1.2.1/lib/statsd.rb:232:in `time'
/home/realvolve/app/vendor/bundle/ruby/2.1.0/gems/sidekiq-pro-1.7.4/lib/sidekiq/middleware/server/statsd.rb:30:in `call'
/home/realvolve/app/vendor/bundle/ruby/2.1.0/gems/sidekiq-3.2.1/lib/sidekiq/middleware/chain.rb:124:in `block in invoke'
/home/realvolve/app/vendor/bundle/ruby/2.1.0/gems/newrelic_rpm-3.6.6.147/lib/new_relic/agent/instrumentation/sidekiq.rb:25:in `block in call'
/home/realvolve/app/vendor/bundle/ruby/2.1.0/gems/newrelic_rpm-3.6.6.147/lib/new_relic/agent/instrumentation/controller_instrumentation.rb:318:in `perform_action_with_newrelic_trace'
/home/realvolve/app/vendor/bundle/ruby/2.1.0/gems/newrelic_rpm-3.6.6.147/lib/new_relic/agent/instrumentation/sidekiq.rb:21:in `call'
/home/realvolve/app/vendor/bundle/ruby/2.1.0/gems/sidekiq-3.2.1/lib/sidekiq/middleware/chain.rb:124:in `block in invoke'
/home/realvolve/app/vendor/bundle/ruby/2.1.0/gems/sidekiq-unique-jobs-3.0.2/lib/sidekiq-unique-jobs/middleware/server/unique_jobs.rb:16:in `call'
/home/realvolve/app/vendor/bundle/ruby/2.1.0/gems/sidekiq-3.2.1/lib/sidekiq/middleware/chain.rb:124:in `block in invoke'
/home/realvolve/app/vendor/bundle/ruby/2.1.0/gems/sidekiq-pro-1.7.4/lib/sidekiq/batch/middleware.rb:26:in `call'
/home/realvolve/app/vendor/bundle/ruby/2.1.0/gems/sidekiq-3.2.1/lib/sidekiq/middleware/chain.rb:124:in `block in invoke'
/home/realvolve/app/vendor/bundle/ruby/2.1.0/gems/sidekiq-3.2.1/lib/sidekiq/middleware/server/active_record.rb:6:in `call'
/home/realvolve/app/vendor/bundle/ruby/2.1.0/gems/sidekiq-3.2.1/lib/sidekiq/middleware/chain.rb:124:in `block in invoke'
/home/realvolve/app/vendor/bundle/ruby/2.1.0/gems/sidekiq-3.2.1/lib/sidekiq/middleware/server/retry_jobs.rb:62:in `call'
/home/realvolve/app/vendor/bundle/ruby/2.1.0/gems/sidekiq-3.2.1/lib/sidekiq/middleware/chain.rb:124:in `block in invoke'
/home/realvolve/app/vendor/bundle/ruby/2.1.0/gems/sidekiq-failures-0.4.3/lib/sidekiq/failures/middleware.rb:9:in `call'
/home/realvolve/app/vendor/bundle/ruby/2.1.0/gems/sidekiq-3.2.1/lib/sidekiq/middleware/chain.rb:124:in `block in invoke'
/home/realvolve/app/vendor/bundle/ruby/2.1.0/gems/sidekiq-3.2.1/lib/sidekiq/middleware/server/logging.rb:11:in `block in call'
/home/realvolve/app/vendor/bundle/ruby/2.1.0/gems/sidekiq-3.2.1/lib/sidekiq/logging.rb:22:in `with_context'
/home/realvolve/app/vendor/bundle/ruby/2.1.0/gems/sidekiq-3.2.1/lib/sidekiq/middleware/server/logging.rb:7:in `call'
/home/realvolve/app/vendor/bundle/ruby/2.1.0/gems/sidekiq-3.2.1/lib/sidekiq/middleware/chain.rb:124:in `block in invoke'
/home/realvolve/app/vendor/bundle/ruby/2.1.0/gems/sidekiq-3.2.1/lib/sidekiq/middleware/chain.rb:127:in `call'
/home/realvolve/app/vendor/bundle/ruby/2.1.0/gems/sidekiq-3.2.1/lib/sidekiq/middleware/chain.rb:127:in `invoke'
/home/realvolve/app/vendor/bundle/ruby/2.1.0/gems/sidekiq-3.2.1/lib/sidekiq/processor.rb:51:in `block in process'
/home/realvolve/app/vendor/bundle/ruby/2.1.0/gems/sidekiq-3.2.1/lib/sidekiq/processor.rb:94:in `stats'
/home/realvolve/app/vendor/bundle/ruby/2.1.0/gems/sidekiq-3.2.1/lib/sidekiq/processor.rb:50:in `process'
/home/realvolve/app/vendor/bundle/ruby/2.1.0/gems/celluloid-0.15.2/lib/celluloid/calls.rb:25:in `public_send'
/home/realvolve/app/vendor/bundle/ruby/2.1.0/gems/celluloid-0.15.2/lib/celluloid/calls.rb:25:in `dispatch'
/home/realvolve/app/vendor/bundle/ruby/2.1.0/gems/celluloid-0.15.2/lib/celluloid/calls.rb:122:in `dispatch'
/home/realvolve/app/vendor/bundle/ruby/2.1.0/gems/celluloid-0.15.2/lib/celluloid/actor.rb:322:in `block in handle_message'
/home/realvolve/app/vendor/bundle/ruby/2.1.0/gems/celluloid-0.15.2/lib/celluloid/actor.rb:416:in `block in task'
/home/realvolve/app/vendor/bundle/ruby/2.1.0/gems/celluloid-0.15.2/lib/celluloid/tasks.rb:55:in `block in initialize'
/home/realvolve/app/vendor/bundle/ruby/2.1.0/gems/celluloid-0.15.2/lib/celluloid/tasks/task_fiber.rb:13:in `block in create'
```

**Comment by mperham** [comment]
That blog is 2.5 years old and only applies to Sidekiq 2. Your trace looks normal, it just looks like a long running job.

**Comment by abhijitsinha** [comment]
Thanks @mperham for the response. But this is running since 4 days now whereas it should not same job with different params runs in less than a second or two. Is there a way to kill this job?

**Comment by mperham** [comment]
Restart the process.

**Comment by abhijitsinha** [comment]
@mperham Sure, thanks. We tried restarting the process but it did not work. 

Let me try again and get back. Closing the issue for now.

**Comment by abhijitsinha** [comment]
Hello @mperham - we restarted the process.

The job is back again and running endlessly. This is very strange.

**Comment by mperham** [comment]
Are you using reliable fetch?  Is it possible that's a job which is stuck in an infinite loop?  Is the machine showing >= 100% CPU usage?

**Comment by abhijitsinha** [comment]
yes we are using reliable fetch. CPU usage is normal. Its very weird. 

How do we inspect the parameters of job which is already in `busy`? Can we check redis? 

I know we can check it from sidekiq UI but the jobs params are lengthy so it do not appear completely. It is truncated.

**Comment by abhijitsinha** [comment]
Hello @mperham 

 Is the machine showing >= 100% CPU usage?

What will this lead to if the CPU usage is more than 100% - other jobs run normal though few get into an infinite loop.

**Comment by mperham** [comment]
Yes, the job is in Redis, in the private queue associated with the Sidekiq process.  The queue is named "queue:#{queue}_#{hostname}_#{index}".  You can use `redis-cli keys queue:*` to get the exact name.

100% means you are pegging a core and in an infinite loop.

**Comment by abhijitsinha** [comment]
Thanks @mperham - we were able to remove that job from redis.

Couple of queries - 
1. Is there a way to monitor such jobs consuming large memory and stuck in infinite loop.
2. Can we kill/remove such jobs when memory consumption is high automatically.

Please let us know if you have any suggestion around this. Thanks

**Comment by mperham** [comment]
Both are possible if you want to invest the time but I don't have any specific suggestions.  You might look into using [inspeqtor](https://github.com/mperham/inspeqtor) to monitor the process memory or CPU.

**Comment by abhijitsinha** [comment]
@mperham Thanks for the all the help. Yes, we have inspeqtor in place to monitor those aspects.

Sharing below note to help others  -

We are using the below code to identify the queue name on production redis machine.

```
redis = Redis.current

redis.scan_each(match: 'queue:*') do |queue_key_name|
  puts queue_key_name
end
```

Once you have queue_name - we can use the code below - please note - the logic to match the code can vary on how precisely you want to filter a job in case you have many jobs having similar parameters - 

```
Sidekiq.redis do |r|
   queue_key = "queue_name" #queue_name_obtained_from_step_1
   @jobs = r.lrange(queue_key, 0, -1)
   @jobs.each do |job|
     r.lrem(queue_key, 0, job) if JSON.load(job)['args'].include?(109857)
   end
 end
```

**Comment by sampatbadhe** [comment]
:+1: @abhijitsinha

**Comment by tjoyal** [comment]
Late to the party, but might as well update this. 

You reference this line in v1.8.2
https://github.com/premailer/premailer/blob/265428a59953bc87f19d6b9554b9db8c5608a300/lib/premailer/html_to_plain_text.rb#L29

A fix was merged: https://github.com/premailer/premailer/pull/346
